File: src/tqi/scoring/ptal.py
"""PTAL (Public Transport Accessibility Level) computation.

Reference: Transport for London PTAL methodology.
Originally developed by London Borough of Hammersmith and Fulham (1992),
now standard across London and adopted internationally.
"""

from math import cos, radians

import numpy as np
from scipy.spatial import cKDTree

from tqi.config import (
    EARTH_RADIUS_M,
    PTAL_BUS_CATCHMENT_M,
    PTAL_GRADES,
    PTAL_WALK_SPEED_M_PER_MIN,
)
from tqi.gtfs.parse import GTFSFeed


def ptal_grade(ai: float) -> str:
    """Map an Accessibility Index value to a PTAL grade (1a–6b)."""
    for max_ai, grade in PTAL_GRADES:
        if ai <= max_ai:
            return grade
    return "6b"


def compute_ptal(
    grid: np.ndarray,
    feed: GTFSFeed,
) -> tuple[np.ndarray, np.ndarray]:
    """Compute PTAL Accessibility Index for each grid point.

    Args:
        grid: (N, 2) array of [lat, lon].
        feed: Filtered GTFSFeed with stop_times, trips, routes, stops.

    Returns:
        ai_values: (N,) array of Accessibility Index values.
        grades: (N,) array of PTAL grade strings.
    """
    n_points = len(grid)

    # Build stop spatial index
    stop_lats = feed.stops["stop_lat"].values.astype(float)
    stop_lons = feed.stops["stop_lon"].values.astype(float)
    stop_ids = feed.stops["stop_id"].values

    center_lat_rad = radians(float(stop_lats.mean()))
    stop_xy = np.column_stack([
        np.radians(stop_lons) * EARTH_RADIUS_M * cos(center_lat_rad),
        np.radians(stop_lats) * EARTH_RADIUS_M,
    ])
    tree = cKDTree(stop_xy)

    # Project grid
    grid_xy = np.column_stack([
        np.radians(grid[:, 1]) * EARTH_RADIUS_M * cos(center_lat_rad),
        np.radians(grid[:, 0]) * EARTH_RADIUS_M,
    ])

    # Precompute route headways at each stop
    # For each (stop_id, route_short_name): median headway in minutes
    st = feed.stop_times.merge(feed.trips[["trip_id", "route_id"]], on="trip_id")
    st = st.merge(feed.routes[["route_id", "route_short_name"]], on="route_id")

    stop_route_headways: dict[tuple[str, str], float] = {}

    for (stop_id, route_name), group in st.groupby(["stop_id", "route_short_name"]):
        dep_times = sorted(group["departure_min"].unique())
        if len(dep_times) < 2:
            stop_route_headways[(stop_id, route_name)] = 120.0  # single trip, very long headway
            continue
        headways = np.diff(dep_times)
        headways = headways[headways > 0]
        if len(headways) > 0:
            stop_route_headways[(stop_id, route_name)] = float(np.median(headways))
        else:
            stop_route_headways[(stop_id, route_name)] = 120.0

    # Build stop_id → list of (route_name, headway) lookup
    stop_routes: dict[str, list[tuple[str, float]]] = {}
    for (stop_id, route_name), hw in stop_route_headways.items():
        if stop_id not in stop_routes:
            stop_routes[stop_id] = []
        stop_routes[stop_id].append((route_name, hw))

    # Compute PTAL for each grid point
    ai_values = np.zeros(n_points)

    for i in range(n_points):
        nearby_idx = tree.query_ball_point(grid_xy[i], r=PTAL_BUS_CATCHMENT_M)
        if not nearby_idx:
            continue

        # Collect all (route_name, EDF) pairs from all nearby stops
        route_best_edf: dict[str, float] = {}

        for si in nearby_idx:
            sid = stop_ids[si]
            if sid not in stop_routes:
                continue

            # Walk distance and time
            dx = stop_xy[si, 0] - grid_xy[i, 0]
            dy = stop_xy[si, 1] - grid_xy[i, 1]
            dist_m = (dx**2 + dy**2) ** 0.5
            walk_min = dist_m / PTAL_WALK_SPEED_M_PER_MIN

            for route_name, headway in stop_routes[sid]:
                # Total access time = walk time + average wait (half headway)
                total_access = walk_min + headway / 2.0
                if total_access <= 0:
                    continue
                # Equivalent Doorstep Frequency
                edf = 30.0 / total_access
                # Keep best EDF for each route (closest stop / best headway)
                if route_name not in route_best_edf or edf > route_best_edf[route_name]:
                    route_best_edf[route_name] = edf

        if not route_best_edf:
            continue

        # PTAL weighting: best route = 1.0, all others = 0.5
        edfs = sorted(route_best_edf.values(), reverse=True)
        ai = edfs[0]  # best route, weight 1.0
        for edf in edfs[1:]:
            ai += 0.5 * edf  # other routes, weight 0.5

        ai_values[i] = ai

    # Map to grades
    grades = np.array([ptal_grade(ai) for ai in ai_values])

    return ai_values, grades

File: go/internal/scoring/ptal.go
package scoring

import (
	"math"
	"sort"

	"github.com/jasondyck/chwk-tqi/internal/config"
	"github.com/jasondyck/chwk-tqi/internal/geo"
	"github.com/jasondyck/chwk-tqi/internal/grid"
	"github.com/jasondyck/chwk-tqi/internal/gtfs"
)

// PTALResult holds PTAL accessibility indices and grades for a set of points.
type PTALResult struct {
	Values []float64
	Grades []string
}

// PTALGradeFromAI maps an accessibility index to a PTAL grade string.
func PTALGradeFromAI(ai float64) string {
	for _, g := range config.PTALGrades {
		if ai <= g.MaxAI {
			return g.Grade
		}
	}
	return "6b"
}

// ComputePTAL calculates PTAL using the TfL methodology.
// For each grid point, find stops within 640 m, compute EDF per (stop, route),
// then AI = bestEDF + 0.5 * sum(otherEDFs).
func ComputePTAL(points []grid.Point, feed *gtfs.Feed) *PTALResult {
	// Build trip -> route mapping.
	tripRoute := make(map[string]string, len(feed.Trips))
	for _, tr := range feed.Trips {
		tripRoute[tr.TripID] = tr.RouteID
	}

	// Collect first departure per trip.
	tripFirstDep := make(map[string]int)
	for _, st := range feed.StopTimes {
		if cur, ok := tripFirstDep[st.TripID]; !ok || st.DepartureMin < cur {
			tripFirstDep[st.TripID] = st.DepartureMin
		}
	}

	// Build stop -> set of routes and per-(stop, route) headways.
	type stopRouteKey struct{ stopID, routeID string }
	srTrips := make(map[stopRouteKey][]int) // departure minutes

	// Collect stop-level trips: for each stop_time, note which route serves that stop.
	stopTrips := make(map[string]map[string]bool) // stopID -> set of tripIDs
	for _, st := range feed.StopTimes {
		if stopTrips[st.StopID] == nil {
			stopTrips[st.StopID] = make(map[string]bool)
		}
		stopTrips[st.StopID][st.TripID] = true
	}

	for stopID, trips := range stopTrips {
		for tripID := range trips {
			routeID, ok := tripRoute[tripID]
			if !ok {
				continue
			}
			dep, ok := tripFirstDep[tripID]
			if !ok {
				continue
			}
			k := stopRouteKey{stopID, routeID}
			srTrips[k] = append(srTrips[k], dep)
		}
	}

	// Compute median headway per (stop, route).
	srHeadway := make(map[stopRouteKey]float64)
	for k, deps := range srTrips {
		if len(deps) < 2 {
			srHeadway[k] = 60.0 // default if only one trip
			continue
		}
		sort.Ints(deps)
		var headways []float64
		for i := 1; i < len(deps); i++ {
			hw := float64(deps[i] - deps[i-1])
			if hw > 0 {
				headways = append(headways, hw)
			}
		}
		if len(headways) == 0 {
			srHeadway[k] = 60.0
		} else {
			sort.Float64s(headways)
			srHeadway[k] = medianFloat(headways)
		}
	}

	// Build stop index.
	stopByID := make(map[string]*gtfs.Stop, len(feed.Stops))
	for i := range feed.Stops {
		stopByID[feed.Stops[i].StopID] = &feed.Stops[i]
	}

	// Collect unique routes per stop.
	stopRoutes := make(map[string]map[string]bool)
	for k := range srHeadway {
		if stopRoutes[k.stopID] == nil {
			stopRoutes[k.stopID] = make(map[string]bool)
		}
		stopRoutes[k.stopID][k.routeID] = true
	}

	result := &PTALResult{
		Values: make([]float64, len(points)),
		Grades: make([]string, len(points)),
	}

	for i, pt := range points {
		var edfs []float64

		for _, stop := range feed.Stops {
			distM := geo.Haversine(pt.Lat, pt.Lon, stop.StopLat, stop.StopLon) * 1000.0
			if distM > float64(config.PTALBusCatchmentM) {
				continue
			}

			walkTimeMin := distM / config.PTALWalkSpeedMPerMin

			routes, ok := stopRoutes[stop.StopID]
			if !ok {
				continue
			}

			for routeID := range routes {
				k := stopRouteKey{stop.StopID, routeID}
				hw, ok := srHeadway[k]
				if !ok {
					continue
				}
				edf := 30.0 / (walkTimeMin + hw/2.0)
				edfs = append(edfs, edf)
			}
		}

		if len(edfs) == 0 {
			result.Values[i] = 0
			result.Grades[i] = PTALGradeFromAI(0)
			continue
		}

		// Sort descending to get best EDF first.
		sort.Sort(sort.Reverse(sort.Float64Slice(edfs)))

		ai := edfs[0]
		for j := 1; j < len(edfs); j++ {
			ai += 0.5 * edfs[j]
		}

		if math.IsNaN(ai) || math.IsInf(ai, 0) {
			ai = 0
		}

		result.Values[i] = ai
		result.Grades[i] = PTALGradeFromAI(ai)
	}

	return result
}

func medianFloat(sorted []float64) float64 {
	n := len(sorted)
	if n == 0 {
		return 0
	}
	if n%2 == 0 {
		return (sorted[n/2-1] + sorted[n/2]) / 2.0
	}
	return sorted[n/2]
}
